Update installation instructions, add Spack env file, upgrade container

nilsvu opened the pull request:
## Proposed changes

- Move instructions on building Charm++ and SpECTRE to separate sections so they aren't scattered in the Docker/Singularity/Spack sections.
- Fix broken links, the Charm++ target to compile, etc.
- Add Spack env file to help installing dependencies in a Spack environment.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
To compile the Doxygen documentation, upgrade to the recently released bugfix version 1.9.3. Earlier versions also work, but the documentation may not look the same as on https://spectre-code.org.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add Spack environment file

nilsvu committed:
Update, correct and reformulate some installation instructions

- Move instructions on building Charm++ and SpECTRE to separate sections
  so they aren't scattered in the Docker/Singularity/Spack sections.
- Fix broken links, the Charm++ target to compile, etc.

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -327,77 +282,71 @@ yaml-cpp@develop~shared` (note that is a tilde and not a dash in front
+  `multicore-linux-x86_64`, or `multicore-darwin-x86_64` on macOS. On an HPC
+  system the correct Charm++ target architecture depends on the machine's
+  inter-node communication architecture.
+- Both Charm++ and SpECTRE must be compiled using the same compiler,
```
This is not try, the compilers just need to be ABI compatible. I only have a GCC build of Charm++ but use Clang for almost all spectre things.

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -327,77 +282,71 @@ yaml-cpp@develop~shared` (note that is a tilde and not a dash in front
+- On a personal machine the correct target architecture is likely
+  `multicore-linux-x86_64`, or `multicore-darwin-x86_64` on macOS. On an HPC
+  system the correct Charm++ target architecture depends on the machine's
+  inter-node communication architecture.
```
Maybe add "It might take some experimenting to figure out which Charm++ configuration provides the best performance."

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -327,77 +282,71 @@ yaml-cpp@develop~shared` (note that is a tilde and not a dash in front
+1. Create a build directory where you would like to compile SpECTRE. In the
+   Docker container you could create, e.g., `/work/spectre-build`. It can be
+   useful to add a timestamp or descriptive label to the name of the build
+   directory since you may create more later, e.g., `build_YYYY-MM-DD` or
```
let's not endorse version/date numbering directories... That's literally what version control was created to eliminate. I think if anything, having `build-compiler-mode` like you have is all we should say.

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -327,77 +282,71 @@ yaml-cpp@develop~shared` (note that is a tilde and not a dash in front
+   See \ref common_cmake_flags for documentation on possible configuration
+   options.
+4. When cmake configuration is done, you are ready to build target executables.
+   Compile unit tests with `make unit-tests -jN` where `N` is the number of
```
"If you are using make, compile unit tests..."

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -327,77 +282,71 @@ yaml-cpp@develop~shared` (note that is a tilde and not a dash in front
+   Compile unit tests with `make unit-tests -jN` where `N` is the number of
+   cores to build on in parallel (e.g. `-j4`).
+
+   - You can see the list of available targets by running `make list`.
```
"or by using tab completion. Note that the Ninja generator allows you to compile individual source files too."

nilsvu replied:
Reworded this a bit, so pushed a fixup

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -310,20 +310,13 @@ To use modules with Spack, enable Spack's shell support by adding
-You can also install CMake, OpenBLAS, Boost, and HDF5 from Spack.
-To load the packages you've installed from Spack run `spack load PACKAGE`,
-or (equivalently) use the `module load` command.
+install the SpECTRE dependencies using Spack. A convenient way to install the
```
"A convenient" -> "One"

Or maybe say "SpECTRE provides a [Spack environment]..."

nilsdeppe reviewed (changes requested):
Thanks for doing this! Please go ahead and squash directly

nilsvu committed:
Upgrade Doxygen in container

A bug was fixed in this release.

nilsvu commented:
Also added a commit to upgrade Doxygen in the container to their latest bugfix release.

Edit: .. and another to fix a broken download link for Intel SDE in the container.

nilsvu committed:
Upgrade Intel SDE in container

The old version was unavailable to download for some reason. The new
version seems to work.

nilsdeppe reviewed (commented):
LGTM. Should there be upgrade instructions in the PR description saying version 1.9.3 of Doxygen is needed?

nilsvu commented:
Sure, added upgrade instructions. Also pushed the container.

nilsvu commented:
@nilsdeppe pushed another fixup where I collected notes for running Spack, and replaced some instructions with links to external docs.

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -237,52 +237,50 @@ To build SpECTRE with Singularity you must:
-dependencies is with a [Spack environment](https://spack.readthedocs.io/en/latest/environments.html):
+[Install Spack](https://spack.readthedocs.io/en/latest/getting_started.html) by
+cloning it into `SPACK_DIR` (a directory of your choice). For example, you can
+clone Spack into `SPECTRE_ROOT/support/spack`. Then, enable Spack's shell
```
Cloning into the spectre source tree seems like a recipe for people to accidentally try to merge Spack into SpECTRE as part of a PR. I don't think it's good to recommend this location.

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -237,52 +237,50 @@ To build SpECTRE with Singularity you must:
+cloning it into `SPACK_DIR` (a directory of your choice). For example, you can
+clone Spack into `SPECTRE_ROOT/support/spack`. Then, enable Spack's shell
+support with `source SPACK_DIR/share/spack/setup-env.sh`. Consider adding this
+line to your `.bash_profile` or `.bashrc`. Refer to [Spack's getting started
```
maybe add: "(or `.zshrc` if you are using ZSH like on macOS)"

nilsvu replied:
Changed to: Consider adding this line to your `.bash_profile`, `.bashrc`, or similar.

nilsdeppe commented on docs/Installation/Installation.md:
```diff
@@ -237,52 +237,50 @@ To build SpECTRE with Singularity you must:
+- To avoid reinstalling lots of system-provided packages with Spack, use the
+  `spack external find` feature and the `--reuse` flag to `spack concretize` (or
+  `spack install`). You can also install some of the dependencies with your
+  system's package manager in advance, e.g., with `apt-get` or `brew`. If they
```
I think Debian now recommends using `apt` instead of `apt-get` (though I don't know details)

nilsdeppe reviewed (commented):
LGTM, few small suggestion you can squash immediately

nilsvu commented:
done @nilsdeppe

nilsdeppe reviewed (approved):
LGTM! Thanks for cleaning this up :D

nilsvu commented:
@wthrowe @kidder could one of you take a look at this, or merge directly?

wthrowe reviewed (approved):
Test failure is a bunch of timeouts.  Ignoring.
